src/riak_kv_vnode_status_mgr.erl

```diff
@@ -0,0 +1,408 @@
+                         {reply, {ok, {VnodeId :: binary(),
+                                       Counter :: non_neg_integer(),
+                                       LeaseTo :: non_neg_integer()}},
+                          #state{}}.
+handle_call({vnodeid, LeaseSize}, _From, State) ->
+    #state{status_file=File} = State,
+    {ok, Status} = read_vnode_status(File),
+    %% Note: this is subtle change to this function, now it will
+    %% _always_ trigger a store of the new status, since the lease
+    %% will always be moving upwards. A vnode that starts, and
+    %% crashes, and starts, and crashes over and over will burn
+    %% through a lot of counter (or vnode ids (if the leases are very
+    %% large.))
+    {Counter, LeaseTo, VnodeId, Status2} = get_counter_lease(LeaseSize, Status),
+    ok = write_vnode_status(Status2, File),
+    Res = {ok, {VnodeId, Counter, LeaseTo}},
+    {reply, Res, State};
+handle_call(clear, _From, State) ->
+    #state{status_file=File} = State,
+    {ok, Status} = read_vnode_status(File),
+    Status2 = orddict:erase(counter, orddict:erase(vnodeid, Status)),
+    ok = write_vnode_status(Status2, File),
+    {reply, {ok, cleared}, State};
+handle_call(status, _From, State) ->
+    #state{status_file=File} = State,
+    {ok, Status} = read_vnode_status(File),
+    {reply, {ok, Status}, State};
+handle_call(stop, _From, State) ->
+    {stop, normal, ok, State}.
+
+%%--------------------------------------------------------------------
+%% @private
+%%--------------------------------------------------------------------
+
+-spec handle_cast({lease, non_neg_integer()}, #state{}) ->
+                         {noreply, #state{}}.
+handle_cast({lease, LeaseSize}, State) ->
+    #state{status_file=File, vnode_pid=Pid} = State,
+    {ok, Status} = read_vnode_status(File),
+    {_Counter, LeaseTo, VnodeId, UpdStatus} = get_counter_lease(LeaseSize, Status),
+    ok = write_vnode_status(UpdStatus, File),
+    Pid ! {counter_lease, {self(), VnodeId, LeaseTo}},
+    {noreply, State}.
+
+handle_info(_Info, State) ->
+    {noreply, State}.
+
+terminate(_Reason, _State) ->
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+%% @private monotonically advance the counter lease. Guarded at
+%% interface to server.
+-spec get_counter_lease(non_neg_integer(), status()) ->
+                           {non_neg_integer(), non_neg_integer(), binary(), status()}.
+get_counter_lease(LeaseSize, Status) ->
+    PrevLease = get_status_item(counter, Status, undefined),
+    VnodeId0 = get_status_item(vnodeid, Status, undefined),
+    Version = get_status_item(version, Status, 1),
+
+    case {Version, PrevLease, VnodeId0} of
+        {_, _, undefined} ->
+            new_id_and_counter(Status, min(LeaseSize, ?MAX_CNTR));
```

Comment:
Ooooh, good one, because actually the `LeaseSize` isn't normalised in the final case clause's body.